src/ucp/core/ucp_rkey.c

```diff
@@ -1183,54 +1183,66 @@ void ucp_rkey_proto_select_dump(ucp_worker_h worker,
 }
 
 ucs_status_t
-ucp_rkey_compare(ucp_context_h context, ucp_rkey_h rkey1, ucp_rkey_h rkey2,
+ucp_rkey_compare(ucp_worker_h worker, ucp_rkey_h rkey1, ucp_rkey_h rkey2,
                  const ucp_rkey_compare_params_t *params, int *result)
 {
-    uct_rkey_compare_params_t uct_params = {};
-    ucs_status_t status                  = UCS_OK;
+    ucs_status_t status;
+    uct_rkey_compare_params_t uct_params;
     uct_component_h cmpt;
     ucp_md_index_t remote_md_index;
-    unsigned i;
+    unsigned rkey_index;
     uct_rkey_t uct_rkey1, uct_rkey2;
-    ucp_worker_cfg_index_t cfg_index1, cfg_index2;
 
     if ((params->field_mask != 0) || (result == NULL)) {
+        ucs_error("invalid field_mask 0x%" PRIu64 " or null result passed",
+                  params->field_mask);
         return UCS_ERR_INVALID_PARAM;
     }
 
     /* Matching config indices means that the possibly unrelated remote MDs all
      * resolve to the same local components.
      */
-    if (context->config.ext.proto_enable) {
-        cfg_index1 = rkey1->cfg_index;
-        cfg_index2 = rkey2->cfg_index;
+    if (worker->context->config.ext.proto_enable) {
+        if (rkey1->cfg_index != rkey2->cfg_index) {
+            *result = (rkey1->cfg_index > rkey2->cfg_index) ? 1 : -1;
+            return UCS_OK;
+        }
+
+        ucs_assertv(rkey1->md_map == rkey2->md_map,
+                    "rkey1->md_map=0x%" PRIx64 " rkey2->md_map=0x%" PRIx64,
+                    rkey1->md_map, rkey2->md_map);
     } else {
-        cfg_index1 = rkey1->cache.ep_cfg_index;
-        cfg_index2 = rkey2->cache.ep_cfg_index;
-    }
+        if (rkey1->cache.ep_cfg_index != rkey2->cache.ep_cfg_index) {
+            *result = (rkey1->cache.ep_cfg_index > rkey2->cache.ep_cfg_index) ?
+                              1 :
+                              -1;
+            return UCS_OK;
+        }
 
-    if (cfg_index1 != cfg_index2) {
-        *result = cfg_index1 > cfg_index2 ? 1 : -1;
-        return UCS_OK;
+        if (rkey1->md_map != rkey2->md_map) {
+            *result = (rkey1->md_map > rkey2->md_map) ? 1 : -1;
+            return UCS_OK;
+        }
```

Comment:
maybe move this part outside of if/else? then could simplify the code by also moving the if conditions at lines 1206,1215 outside of the if/ese

Reply:
tried to address comment, pls double check